URGENT: Check if first requests are sent without authentication token

I was reading the implementation of Authenticator of OkHTTP3 and it seems it only calls the callback to set authentication headers IF the server fails and give 401. Please check if this is true since this can cause unauthenticated information to be shown to the user for endpoints which support both authenticated and unauthenticated users

@liveHarshit @GOVINDDIXIT Can you please check this?

> What error?

```
E/KOIN: [ERROR] - Error while resolving instance for class 'okhttp3.Authenticator' - error: org.koin.error.NoBeanDefFoundException: No compatible definition found. Check your module definition 
E/KOIN: [ERROR] - Error while resolving instance for class 'okhttp3.OkHttpClient' - error: org.koin.error.BeanInstanceCreationException: Can't create definition for 'Single [name='OkHttpClient',class='okhttp3.OkHttpClient']' due to error :
    		No compatible definition found. Check your module definition
    		org.koin.core.bean.BeanRegistry.checkedResult(BeanRegistry.kt:120)
    		org.koin.core.bean.BeanRegistry.retrieveDefinition(BeanRegistry.kt:87)
```

> There's no such need, just declare it in Modules.kt

Yes I have done it in network module - `single { RequestAuthenticator(get()) as Interceptor }` but after running the app - 


> E/KOIN: [ERROR] - Error while resolving instance for class 'okhttp3.Authenticator' - error: org.koin.error.NoBeanDefFoundException: No compatible definition found. Check your module definition E/KOIN: [ERROR] - Error while resolving instance for class 'okhttp3.OkHttpClient' - error: org.koin.error.BeanInstanceCreationException: Can't create definition for 'Single [name='OkHttpClient',class='okhttp3.OkHttpClient']' due to error : No compatible definition found. Check your module definition org.koin.core.bean.BeanRegistry.checkedResult(BeanRegistry.kt:120) org.koin.core.bean.BeanRegistry.retrieveDefinition(BeanRegistry.kt:87)

--- a/app/src/main/java/org/fossasia/openevent/general/auth/RequestAuthenticator.kt
+++ b/app/src/main/java/org/fossasia/openevent/general/auth/RequestAuthenticator.kt
@@ -1,18 +1,19 @@
 package org.fossasia.openevent.general.auth
 
-import okhttp3.Authenticator
-import okhttp3.Request
+import okhttp3.Interceptor
 import okhttp3.Response
-import okhttp3.Route
 
-class RequestAuthenticator(private val authHolder: AuthHolder) : Authenticator {
+class RequestAuthenticator(private val authHolder: AuthHolder) : Interceptor {
 
-    override fun authenticate(route: Route?, response: Response): Request? {
-        val authorization: String? = authHolder.getAuthorization()
-        return if (response.request().header("Authorization") == null && authorization != null) {
-            response.request().newBuilder()
-                    .header("Authorization", authorization)
-                    .build()
-        } else null // Give up, we already tried authenticating
+    override fun intercept(chain: Interceptor.Chain): Response? {
+        val authorization = authHolder.getAuthorization()
+        val original = chain.request()
+        return if (authorization != null) {
+            val request = original.newBuilder()
+                .header("Authorization", authorization)
+                .build()
+            chain.proceed(request)
+        } else
+            chain.proceed(original)
     }
 }

--- a/app/src/main/java/org/fossasia/openevent/general/di/Modules.kt
+++ b/app/src/main/java/org/fossasia/openevent/general/di/Modules.kt
@@ -5,7 +5,7 @@ import com.fasterxml.jackson.databind.DeserializationFeature
 import com.fasterxml.jackson.databind.ObjectMapper
 import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
 import com.github.jasminb.jsonapi.retrofit.JSONAPIConverterFactory
-import okhttp3.Authenticator
+import okhttp3.Interceptor
 import okhttp3.OkHttpClient
 import okhttp3.logging.HttpLoggingInterceptor
 import org.fossasia.openevent.general.BuildConfig
@@ -153,7 +153,7 @@ val networkModule = module {
         objectMapper
     }
 
-    single { RequestAuthenticator(get()) as Authenticator }
+    single { RequestAuthenticator(get()) as Interceptor }
 
     single {
         val connectTimeout = 15 // 15s
@@ -165,7 +165,7 @@ val networkModule = module {
                 .addInterceptor(
                         HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY)
                 )
-                .authenticator(get())
+                .addInterceptor(get())
                 .build()
     }
 

